### ISSUE TITLE
Make geometries in GeometryCollection available via getter

### ISSUE THREAD

**Issue by oers** [locationtech/jts#224]
The geometries field is not accessible via getter in GeometryCollection (and mybe use generics here, too).

It would be great to get access to this field.

For my usecase I want to merge two large Multipoints (100k Points each) and the fastest way is by copying the arrays directly.

Currently I use FieldUtils from apache commons to get the field, which I find not very convenient.
```
            Point[] headCoordinates;
            Point[] tailCoordinates;
            try {
                headCoordinates = (Point[])FieldUtils.readField(head.getGeometry(), "geometries", true);
                tailCoordinates = (Point[])FieldUtils.readField(toMerge.getGeometry(), "geometries", true);
            } catch (Exception e) {
                throw new IllegalStateException(e);
            }
            Point[] concat = ObjectArrays.concat(headCoordinates, tailCoordinates, Point.class);
            head.setGeometry(GeometryFunctions.getGeometryFactory(head.getGeometry().getSRID()).createMultiPoint(concat));
```

**Comment by dr-jts** [comment]
The internal representation of GeometryCollection (and all other geometries) is hidden by design.  That encapsulates it and allows change without breaking public API.  Also, exposing arrays means that they can be changed by external code without the parent being aware of it.  And, geometries are meant to be immutable.

For all these reasons, this change will not be implemented as proposed.  Perhaps there is some other way to accomplish this?

**Comment by jnh5y** [comment]
@oers it sounds like you'd really like a method for unioning two multipoints quickly?  Perhaps that's a better issue to work through.

**Comment by dr-jts** [comment]
Good point, @jnh5y .  Although the semantics presented above actually implement a "combine" operation (union would merge identical points).  This could be a method on GeometryCollection, returning an actual GeometryCollection.

It seems like there might be something like that already in JTS - I'll give this some thought.

**Comment by dr-jts** [comment]
There is the `GeometryCombiner `:  https://github.com/locationtech/jts/blob/master/modules/core/src/main/java/org/locationtech/jts/geom/util/GeometryCombiner.java

This is probably not as efficient as the raw array-based approach, however.  Any idea of the performance difference?

**Comment by oers** [comment]
The Geoemtry Combiner needs to iterate over all elements, so it is probably worse the more elements you have opposed to a raw copy in memory.

As for protecting the internal structure:
With the the current architecture it is still possible to change values of any point

```
getGeometryN(1).getCoordinate().x=....
```

So the only thing that is protected by hiding the array is the length of the array. Everything else can be changed at will (which is good, I suppose).
And the missing access to the array leads to having an inefficient GeometryCombiner.

Maybe the combine Method should be implemented in Geometries.

**Comment by dr-jts** [comment]
The ability to change Coordinates is a legacy feature, which is strongly discourage going forward (since it leads to nasty aliasing bugs).  

That said, for performance it's nice to be able to access internal structure within the library itself.  I guess one way to do this is to provide access to a copy of the internal array.  There's always a tension between performance and data encapsulation, and in Java arrays are a prime source of this issue.

The other problem with providing such low-level access is that it constrains future changes to the implementation.  At some point Geometry will become simply an interface, and there will be less appeal to carrying forward a method which returns the component geometries as an array.

Another part of my reluctance is that this is the first time in JTS' 15 year history that this has come up as a requirement.  Is there any other way to optimize your situation?  E.g. manage the Point data externally to JTS, and do the merging there?

**Comment by oers** [comment]
Thank you for your feddback :)

I mostly use jts together with hibernate to save and load geometries. 

MultiPoint etc. are just part of the beans and simple data containers in that sense. So I don't really care about the side effects, because these containers are never displayed on a geotools map or are part of a feature collection / Layer . We use a different approach to visualise the data, due to the performance constraints of Java 2D.

So the ability to access and edit the data directly in Geometries is crucial for fast and easy operations.
I have additional use cases like: deleting a point, moving a point etc. Sometimes  on very large Geometries (several 100k points). 

I could have a look at geolatte, but I am kind of used to jts :)

**Comment by jnh5y** [comment]
@dr-jts as a possibility, I wonder if it'd make sense to expose 'unsafe' methods (or provide a separate 'unsafe' package) to allow users to do things at their own risk?

**Comment by dr-jts** [comment]
@jnh5y Not sure how to implement an "unsafe package"?  Of course can always expose methods and warn users they are breaking encapsulation.  Widening API does come at a long-term cost, however - once they are there it's painful to take them away.

**Comment by dr-jts** [comment]
One thing needed to justify this change is some metrics on how much faster array copy is versus accessor methods.  Is it possible to provide a test harness to demonstrate this?

**Comment by dsmiley** [comment]
> At some point Geometry will become simply an interface, and there will be less appeal to carrying forward a method which returns the component geometries as an array.

Furthermore, at such time users will be able to provide their own implementation such as one that exposes the array.

**Comment by dr-jts** [comment]
@dsmiley Agreed, when Geometry becomes an interface in JTS2 that changes the conversation a bit about what gets exposed where.  

It's still interesting to think about the underlying requirement (i.e. combining Geometry components rather than unioning them) and whether it is a common enough use case that it should be supported by methods that allow efficient implementation.

**Comment by oers** [comment]
Geolatte has a simple solution for this in their GeometryCollection. 
I think this is a good compromise.

```   
/**
     * Returns the components
     *
     * @return an array containing all component objects
     */
    @Override
    @SuppressWarnings("unchecked")
    public G[] components() {
        return (G[]) Arrays.copyOf(this.geometries, this.geometries.length);
    }
```

**Comment by oers** [comment]
Something similar was now done in CoordinateArraySequence.java for 1.6.1

https://github.com/locationtech/jts/blob/534cbb15675be95854c3c6018ee63a1d0c05c533/modules/core/src/main/java/org/locationtech/jts/geom/impl/CoordinateArraySequence.java 

So maybe it is time to allow the same thing for GeometryCollection?
